### PR TITLE
Make fetchGit/fetchTree use Git file ingestion

### PR DESCRIPTION
**Pull request by matthewbauer**
Add support for fetchTree in git objects. Try it with something like:

```
NIX_REMOTE=local nix --store $HOME build --experimental-features 'nix-command flakes' '(builtins.fetchTree { type = "git"; url = "https://github.com/nixos/nix.git"; treeHash = "0674cab54e1226d8e1e1d04572f07c5296edc862"; })'
```

### CODE REVIEW

File: src/libutil/git.cc [hunk 1]
```diff
@@ -154,22 +154,32 @@ GitMode dumpGitBlob(const Path & path, const struct stat st, Sink & sink)
...
+        auto s1 = (format("%d %s") % mode % name).str();
+        std::copy(s1.begin(), s1.end(), std::back_inserter(v1));
+        v1.push_back(0);
+        std::copy(i.second.second.hash, i.second.second.hash + 20, std::back_inserter(v1));
     }
 
-    std::string s2 = (format("tree %d\0%s"s) % s1.size() % s1).str();
+    vector<uint8_t> v2;
+    auto s2 = (format("tree %d"s) % v1.size()).str();
+    std::copy(s2.begin(), s2.end(), std::back_inserter(v2));
```
Comment (Ericson2314) [inline src/libutil/git.cc:177]: Is there a `front_inserter`? Then we could use just one array.

File: src/libutil/git.cc [hunk 2]
```diff
@@ -187,8 +197,17 @@ static GitMode dumpGitInternal(HashType ht, const Path & path, Sink & sink, Path
     else if (S_ISDIR(st.st_mode)) {
         GitTree entries;
         for (auto & i : readDirectory(path))
-            if (filter(path + "/" + i.name))
-                entries[i.name] = dumpGitHashInternal(ht, path + "/" + i.name, filter);
+            if (filter(path + "/" + i.name)) {
+                auto result = dumpGitHashInternal(ht, path + "/" + i.name, filter);
+
+                // correctly observe git order, see
+                // https://github.com/mirage/irmin/issues/352
```
Comment (Ericson2314) [inline src/libutil/git.cc:204]: Nice catch!

File: src/libutil/git.cc (hunk 1 above)
Comment (matthewbauer) [reply src/libutil/git.cc:177]: It looks like there is!

Comment (Ericson2314) [reply src/libutil/git.cc:177]: Oh it doesn't work, because it would be quadratic.

**Pull request merged**